swift/send-message/message-types/twitter/Sources/twitter/send_tweet.swift

```diff
@@ -0,0 +1,78 @@
+import Foundation
+import AsyncHTTPClient
+import NIO
+import NIOFoundationCompat
+import Crypto
+
+class TwitterMessenger : Messenger{
+    private let oauth_consumer_key:String
+    private let oauth_consumer_secret:String
+    private let oauth_token:String
+    private let oauth_token_secret:String
+    private let httpClient: HTTPClient
+
+    init(_ env_vars: [String: String]) throws {
+    guard let oauth_consumer_key = env_vars["TWITTER_API_KEY"],
+        let oauth_consumer_secret = env_vars["TWITTER_API_KEY_SECRET"],
+        let oauth_token = env_vars["TWITTER_ACCESS_TOKEN"],
+        let oauth_token_secret = env_vars["TWITTER_ACCESS_TOKEN_SECRET"] else {
+        throw MessengerError.misconfigurationError(error: "Missing environment variables.")
+    }
+    self.oauth_consumer_key = oauth_consumer_key
+    self.oauth_consumer_secret = oauth_consumer_secret
+    self.oauth_token = oauth_token
+    self.oauth_token_secret = oauth_token_secret
+    httpClient = HTTPClient(eventLoopGroupProvider: .createNew)
+    }
+
+    func sendMessage(messageRequest:Message) async -> Error? {        
+        var tweetText:String
+        if messageRequest.recipient != "" {
+            tweetText = "@" + messageRequest.recipient + " " + messageRequest.content
+        } else {
+            tweetText = messageRequest.content
+        }
+        
+        var request = HTTPClientRequest(url: "https://api.twitter.com/2/tweets")
+        let jsonText : [String:String] = ["text":"\(tweetText)"]
+
+        request.method = .POST
+        request.headers.add(name: "Content-Type", value: "application/json")
+        request.headers.add(name: "Authorization", value: createHeader())
+        
+        do {
+            request.body = .bytes(ByteBuffer(data: (try JSONSerialization.data(withJSONObject: jsonText))))
+            let response = try await httpClient.execute(request, timeout: .seconds(30))
+            try await httpClient.shutdown()
+            if response.headers["location"] == [] {
+                return MessengerError.validationError(error: "Unable to post tweet, API Status Code: \(response.status)")
+            }
+            return nil //Returns no error if tweet was created    
+        } catch {
+            return MessengerError.providerError(error: "Request did not recieve a response or  connection timeout")
+        }
+    } 
+
+    private func createHeader() -> String {
```

Comment:
write a unit test for this function

Reply:
set this function to take function inputs rather than reading directly from the environment, that will make unit testing possible